[autograd] Do not detach when unpacking tensors that do not require grad

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127959
* #129262
* #128545
* #125795

In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops.

soulitzer committed:
Support nested tensor with activation checkpoint

[ghstack-poisoned]

soulitzer committed:
Update on "Support nested tensor with activation checkpoint"




[ghstack-poisoned]

soulitzer committed:
Update on "Support nested tensor with activation checkpoint"




[ghstack-poisoned]

soulitzer committed:
Update on "Support nested tensor with activation checkpoint"




[ghstack-poisoned]

soulitzer committed:
Update on "Support nested tensor with activation checkpoint"




[ghstack-poisoned]

soulitzer committed:
Update on "Support nested tensor with activation checkpoint"




[ghstack-poisoned]

soulitzer committed:
Update on "Support nested tensor with activation checkpoint"




[ghstack-poisoned]

YuqingJ commented on test/test_nestedtensor.py:
```diff
@@ -4496,6 +4501,31 @@ def f(values, offsets):
+            ]
+        )
+
+        def fn(values, offsets):
```
Do you mean def fn(values, lengths) ?

soulitzer replied:
Yup, good catch

soulitzer committed:
Update on "[checkpoint] Activation checkpoint preserves object identity when tensor does not require grad"


This PR ensures that if a tensor does not require grad, and it is saved for backward, (1) unpacking in general preserve the object, (2) non-reentrant checkpoint preserves the object.


[ghstack-poisoned]

soulitzer committed:
Update on "[checkpoint] Activation checkpoint preserves object identity when tensor does not require grad"


This PR ensures that if a tensor does not require grad, and it is saved for backward, (1) unpacking in general preserve the object, (2) non-reentrant checkpoint preserves the object.


[ghstack-poisoned]

soulitzer committed:
Update on "[checkpoint] Activation checkpoint preserves object identity when tensor does not require grad"


This PR ensures that if a tensor does not require grad, and it is saved for backward, (1) unpacking in general preserve the object, (2) non-reentrant checkpoint preserves the object.

For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

soulitzer committed:
Update on "[autograd] Do not detach when unpacking tensors that do not require grad "


In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

soulitzer committed:
Update on "[autograd] Do not detach when unpacking tensors that do not require grad "


In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

soulitzer committed:
Update on "[autograd] Do not detach when unpacking tensors that do not require grad "


In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

YuqingJ commented on test/test_nestedtensor.py:
```diff
@@ -4496,6 +4501,31 @@ def f(values, offsets):
+            nt = convert_jagged_to_nested_tensor(values, offsets, max_length=4)
+            return convert_nt_to_jagged(nt).sum()
+
+        checkpoint(fn, values, lengths, use_reentrant=False, context_fn=context_fn).backward()
```
nit: maybe add another check to make sure values.grad is not None?

soulitzer committed:
Update on "[autograd] Do not detach when unpacking tensors that do not require grad "


In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

soulitzer committed:
Update on "[autograd] Do not detach when unpacking tensors that do not require grad "


In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

soulitzer committed:
Update on "[autograd] Do not detach when unpacking tensors that do not require grad "


In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

soulitzer committed:
Update on "[autograd] Do not detach when unpacking tensors that do not require grad "


In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

soulitzer committed:
Update on "[autograd] Do not detach when unpacking tensors that do not require grad "


In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

soulitzer committed:
Update on "[autograd] Do not detach when unpacking tensors that do not require grad "


In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

soulitzer committed:
Update on "[autograd] Do not detach when unpacking tensors that do not require grad "


In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

soulitzer committed:
Update on "[autograd] Do not detach when unpacking tensors that do not require grad "


In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

soulitzer committed:
Update on "[autograd] Do not detach when unpacking tensors that do not require grad "


In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

soulitzer committed:
Update on "[autograd] Do not detach when unpacking tensors that do not require grad "


In this PR:
- Ensure that if a tensor not requiring grad is saved for backward unpacking does not trigger a detach (unless the user installs a saved tensor pack hook that returns a tensor requiring grad).
- Update non-reentrant checkpoint to also no longer detach for this case.

Alternatives:
- For custom autograd Function, you could directly save on ctx to work around this, but that would not work for when we switch to using custom ops. 


[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

YuqingJ commented:
Ready to merge again?

soulitzer commented:
@pytorchbot merge -i
